### ISSUE TITLE
Python bindings crashing on OSX (installed using Homebrew)

### ISSUE THREAD

**Issue by ricardodeazambuja** [nest/nest-simulator#484]
After installing nest using Homebrew (HEAD or not), every time I open a python shell and type "import nest" it crashes with the message below:
Assertion failed: (out != NULL), function execute, file ../../sli/sli_io.cc, line 401.
Abort trap: 6

OSX 10.10.5 and, according to Homebrew, all dependencies are available:
==> Dependencies
Build: autoconf ✔, automake ✔
Required: libtool ✔, readline ✔
Recommended: gsl ✔

**Comment by heplesser** [comment]
@ricardodeazambuja 
- Can you start the nest binary (should be `/usr/local/bin/nest`)?
- Have you run `brew test nest`? If yes, what result did you get?
- Which Python version are you using, and how have you installed it (system, brew, anaconda, ...?).

**Comment by ricardodeazambuja** [comment]
@heplesser 

I'm using Python 2.7.12, installed from Homebrew (/usr/local/Cellar/python/2.7.12).

I can start the nest binary:
              -- N E S T --

  Copyright (C) 2004 The NEST Initiative
  Version 2.10.0 Sep 17 2016 20:10:47

This program is provided AS IS and comes with
NO WARRANTY. See the file LICENSE for details.

Problems or suggestions?
  Visit http://www.nest-simulator.org

Type 'help' to get more information.
Type 'quit' or CTRL-D to quit NEST.

SLI ] exit

brew install nest --with-python --without-gsl --without-openmp
==> Installing nest from homebrew/science
==> Downloading https://github.com/nest/nest-simulator/releases/download/v2.10.0/nest-2.10.0.tar.gz
Already downloaded: /Users/XXX/Library/Caches/Homebrew/nest-2.10.0.tar.gz
==> Patching
patching file sli/slistartup.cc
==> ../configure --disable-debug --disable-dependency-tracking --prefix=/usr/local/Cellar/nest/2.10.0_2 --wit
==> make
==> make install
🍺  /usr/local/Cellar/nest/2.10.0_2: 4,531 files, 97.8M, built in 2 minutes 3 seconds

$ brew test nest
Testing homebrew/science/nest
==> Using the sandbox
==> /usr/local/Cellar/nest/2.10.0_2/bin/nest --version
==> /usr/local/Cellar/nest/2.10.0_2/share/nest/extras/do_tests.sh --test-pynest
Last 15 lines from /Users/XXX/Library/Logs/Homebrew/nest/test.02.do_tests.sh:
  Not running MPI tests because NEST was compiled without support
  for distributed computing. See the file README for details.

Phase 6: Running PyNEST tests.
------------------------------
Assertion failed: (out != NULL), function execute, file ../../sli/sli_io.cc, line 401.
/usr/local/Cellar/nest/2.10.0_2/share/nest/extras/do_tests.sh: line 604: 31250 Abort trap: 6           "${PYTHON}" "${PYTHON_HARNESS}" >> "${TEST_LOGFILE}"
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory

  PyNEST tests: 
     Passed: 
     Failed: 
/usr/local/Cellar/nest/2.10.0_2/share/nest/extras/do_tests.sh: line 595: 205 +  : syntax error: operand expected (error token is " ")
Error: homebrew/science/nest: failed
Failed executing: /usr/local/Cellar/nest/2.10.0_2/share/nest/extras/do_tests.sh --test-pynest
/usr/local/Library/Homebrew/formula.rb:1608:in `block in system'
/usr/local/Library/Homebrew/formula.rb:1545:in `open'
/usr/local/Library/Homebrew/formula.rb:1545:in `system'
/usr/local/Library/Taps/homebrew/homebrew-science/nest.rb:144:in `block in <class:Nest>'
/usr/local/Library/Homebrew/formula.rb:1446:in `block (2 levels) in run_test'
/usr/local/Library/Homebrew/formula.rb:775:in `with_logging'
/usr/local/Library/Homebrew/formula.rb:1445:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/usr/local/Library/Homebrew/extend/fileutils.rb:74:in `block in run'
/usr/local/Library/Homebrew/extend/fileutils.rb:74:in `chdir'
/usr/local/Library/Homebrew/extend/fileutils.rb:74:in `run'
/usr/local/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:1439:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'

**Comment by tobiastshirt** [comment]
@ricardodeazambuja,

I am having the same issue. Any chance you found a solution?

**Comment by tobiastshirt** [comment]
@heplesser 
(1) nest starts on its own.

(2) Using Python 2.7.12, compiled with brew, on MacOSX 10.10.5

(3) brew test nest comes up with the same error message as the command within python:
Testing homebrew/science/nest
==> Using the sandbox
==> /usr/local/Cellar/nest/2.10.0_3/bin/nest --version
==> /usr/local/Cellar/nest/2.10.0_3/share/nest/extras/do_tests.sh --test-pynest
Last 15 lines from /Users/teichert/Library/Logs/Homebrew/nest/test.02.do_tests.sh:
  Not running MPI tests because NEST was compiled without support
  for distributed computing. See the file README for details.

Phase 6: Running PyNEST tests.
------------------------------
Assertion failed: (out != NULL), function execute, file ../../sli/sli_io.cc, line 401.
/usr/local/Cellar/nest/2.10.0_3/share/nest/extras/do_tests.sh: line 604: 39819 Abort trap: 6           "${PYTHON}" "${PYTHON_HARNESS}" >> "${TEST_LOGFILE}"
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory
cat: pynest_test_numbers.log: No such file or directory

  PyNEST tests: 
     Passed: 
     Failed: 
/usr/local/Cellar/nest/2.10.0_3/share/nest/extras/do_tests.sh: line 595: 205 +  : syntax error: operand expected (error token is " ")
Error: homebrew/science/nest: failed
Failed executing: /usr/local/Cellar/nest/2.10.0_3/share/nest/extras/do_tests.sh --test-pynest
/usr/local/Homebrew/Library/Homebrew/formula.rb:1733:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1671:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1671:in `system'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-science/nest.rb:144:in `block in <class:Nest>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1571:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:832:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1570:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `block in run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `chdir'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1564:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:27:in `<main>'

**Comment by tobiastshirt** [comment]
Update: the issue resolved when adding the following flags to the brew formula in addition to the --with-python flag already set in the previous brew:
--with-mpi --without-gsl --without-openmp --HEAD

the import nest command still fails, but with a different error message

**Comment by heplesser** [comment]
@tobiastshirt You write that the issue resolved with the additional flags, but that you still get an error when running `import nest`. This is confusing: if you mean by "resolved" that all tests pass, then importing nest must work---the Python-based tests all import nest. Could there be a problem with your PYTHONPATH? And what error message are you now getting?

**Comment by tobiastshirt** [comment]
@heplesser 
Thanks for your help, and sorry for being unclear in the last post. When importing nest I now get the following error:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/nest/__init__.py", line 52, in <module>
ImportError: cannot import name pynestkernel

I assume that the original error was resolved because 'brew test nest' now finds an error on Phase 7, rather than Phase 6 as before. 

It could very well be a python path issue. I am new to python so I am not sure how the correct path would look like. Here is the output of python -c 'import sys;print sys.path'. It does include '/usr/local/lib/python2.7/site-packages', which in turn includes a directory names 'nest' and __init__.pyc.

['', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python27.zip', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-darwin', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac/lib-scriptpackages', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-old', '/usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/site-packages', '/Library/Python/2.7/site-packages']

**Comment by heplesser** [comment]
@tobiastshirt Could you try adding `/usr/local/lib/python2.7/site-packages` to your `PYTHONPATH`?  It might be an idea to put that into your `.bashrc`, since it is a quite common path.

**Comment by henriklinden** [comment]
Hi,

I have also run into similar problems when trying to install nest via brew on macOS 10.12. After some helpful advice from @heplesser on the nest_user mailing list I tried a slightly different route: I installed gsl, gcc, cmake and libtool from brew and then compiled the latest nest from git in a python 2.7 conda environement (I did it in the root environment, but a separate environment also works fine) - see below for full installation steps. 

I do get a a lot of compiler warnings (as reported in #486 ) and a few lines of unnessary info on nest startup (same as reported in #415 ) BUT all tests pass and threading also works.

Here is what I did:

0) install homebrew and miniconda2

1) in a terminal type:

brew install gcc gsl cmake libtool

conda install numpy scipy matplotlib ipython cython nose

git clone https://github.com/nest/nest-simulator.git

mkdir nest-build
cd nest-build

cmake -DCMAKE_INSTALL_PREFIX:PATH=$HOME/opt/nest -DCMAKE_C_COMPILER=gcc-6 -DCMAKE_CXX_COMPILER=g++-6 ../nest-simulator/
make
make install
make installcheck

3) add this line to .bash_profile: source /Users/henrik/opt/nest/bin/nest_vars.sh

@heplesser thanks a lot for your help and continuos efforts to make nest work on mac! For some reason the nest_user emails are blocked on my email account by some sort of spam filter (microsoft) so I could not reply to the email you sent (I found it in the nest_user archive).

**Comment by tclose** [comment]
I am having the same problem with the brew version. 

I had a look at that line in sli_io.cc and it relates to the dynamic casting of ostringstream

```
 396 #ifdef HAVE_SSTREAM
 397   std::ostringstream* out = dynamic_cast< std::ostringstream* >( ostreamdatum->get() );
 398 #else
 399   std::ostrstream* out = dynamic_cast< std::ostrstream* >( ostreamdatum->get() );
 400 #endif
 401   assert( out != NULL );
```

I checked the config.log and looks like HAVE_SSTREAM was defined. However, I am guessing that it isn't being consistently defined across the build so that ostreamdatum->get() is returning a std::ostrstream instead of a ostringstream and this is causing the dynamic_cast to fail and return the NULL pointer?

Any ideas on how to debug this further?

**Comment by benoit-girard** [comment]
I am also having the same problem with the brew version. 

I tested the solution proposed by @henriklinden, it seems to work (at least, I can "import nest" in python)

**Comment by heplesser** [comment]
This is appears to be fixed in the current head (see also #636), and documentation has been improved (#615). Therefore, I am closing this Issue.
